File: extra_practice/single_linked.py
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None
    
class SingleLinkedList:
    def __init__(self):
        self.head = Node(None)

    def listPrint(self, head):
        cur = self.head
        idx = 0

        while cur.next != None:
            print(f"list[{idx}] = {cur.value}")
            cur = cur.next
            i += 1
    
    def listSize(self, head) -> int:
        cur = self.head
        size = 0

        while cur.next != None:
            cur = cur.next
            size += 1
        
        return size
    
    def listAt(self, head, pos) -> int:
        cur = self.head
        i = 0

        while i < pos and cur.next != None:
            cur = cur.next
            i += 1

        return cur.next.value

    def listFind(self, head, value) -> int:
        cur = self.head
        i = 0

        while cur.next.value != value and cur.next != None:
            cur = cur.next
            i += 1

        return i
    
    def listInsert(self, head, pos, value):
        cur = self.head
        i = 0

        while i < pos and cur.next != None:
            cur = cur.next
            i += 1

        newNode = Node()
        newNode.value = value

        newNode.next = cur.next
        cur.next = newNode
    
    def listPop(self):
        pass

File: data_structures/single_linked.cpp
// Implementação de uma Single Linked List e funções possíveis

#include <iostream>

using namespace std;

struct Node {
    int val;
    Node* next;
};


void listPrint(Node* head) {
    Node* cur = head;
    int i = 0;

    while (cur->next != nullptr) {
        cout << "list[" << i++ << "]: " << cur->val << endl;
        cur = cur->next;
    }
}

int listSize(Node* head) {
    Node* cur = head;
    int size = 0;

    while (cur->next != nullptr) {
        cur = cur->next;
        size++;
    }

    return size;
}

int listAt(Node* head, int pos) {
    if (pos > listSize(head)) return;

    Node* cur = head;
    int i = 0;

    while (i < pos && cur->next != nullptr) {
        cur = cur->next;
        i++;
    }

    return cur->next->val;
}

int listFind(Node* head, int value) {
    Node* cur = head;
    int i = 0;

    while (cur->next->val != value && cur->next != nullptr) {
        cur = cur->next;
        i++;
    }

    return i;
}

void listInsert(Node* head, int pos, int value) {
    if (pos > listSize(head)) return;

    Node* cur = head;
    int i = 0;

    while (i < pos && cur->next != nullptr) {
        cur = cur->next;
        i++;
    }

    Node* newnode = new Node;
    newnode->val = value;

    newnode->next = cur->next;
    cur->next = newnode;
}

int listPop(Node* head, int pos) {
    if (pos > listSize(head)) return 1;

    Node* cur = head;
    int i = 0;

    while (i < pos && cur->next != nullptr) {
        cur = cur->next;
        i++;
    }

    Node* toDie = cur->next;
    cur->next = toDie->next;
    int deletedValue = toDie->val;
    delete toDie;

    return deletedValue;

}

Node* listNew() {
    Node* head = new Node;
    head->next = nullptr;
    return head;
}

int main(int argc, char** argv) {
    Node* head = listNew();
    int size = stoi(argv[1]);

    for (int v = 0; v < 50; v+=10) {
        listInsert(head, size, v);
        size++;
    }

    listInsert(head, 3, 25);

    listPop(head, 2);

    listPrint(head);

    return 0;
}